lib/linter/apply-disable-directives.js

```diff
@@ -197,62 +197,93 @@ function processUnusedDisableDirectives(allDirectives) {
  * for the exported function, except that `reportUnusedDisableDirectives` is not supported
  * (this function always reports unused disable directives).
  * @returns {{problems: Problem[], unusedDisableDirectives: Problem[]}} An object with a list
- * of filtered problems and unused eslint-disable directives
+ * of problems (including suppressed ones) and unused eslint-disable directives
  */
 function applyDirectives(options) {
     const problems = [];
-    let nextDirectiveIndex = 0;
     let currentGlobalDisableDirective = null;
-    const disabledRuleMap = new Map();
 
-    // enabledRules is only used when there is a current global disable directive.
-    const enabledRules = new Set();
     const usedDisableDirectives = new Set();
 
     for (const problem of options.problems) {
+        let nextDirectiveIndex = 0;
+
+        // isProblemSuppressedBefore indicates whether the problem is suppressed by block directives.
+        const isProblemSuppressedBefore = problem.suppressions ? true : false; // eslint-disable-line no-unneeded-ternary -- problem.suppressions might be changed afterwards.
+
+        const usedDisableDirectivesForProblem = new Set();
```

Comment:
Why do we need a Set of used disable directives for a single problem? There can be only one, and by looking at the code it seems that this Set indeed isn't supposed to have more than one element at any point, so it would greatly simplify the code if we could replace this Set with a reference to one (current) disable directive for the problem.

Reply:
This can be fixed by https://github.com/eslint/eslint/pull/15459#discussion_r785479181